### ISSUE TITLE
Scan-build reports some bugs

### ISSUE THREAD

**Issue by sergiovp** [wazuh/wazuh#4179]
|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.11 | Manager/Agent | Sources | CentOS 7 |

Hello team,

While doing scan-build test on 3.11 branch, it reported some bugs, in both manager and agent.
I've done the test in **CentOS7** and clang's version is **3.4.2**.

+ Agent report:
![image](https://user-images.githubusercontent.com/36703917/67963324-ac26ce00-fbfe-11e9-8056-96ace2d1ae8d.png)

[Scan-Build-Report-Agent.zip](https://github.com/wazuh/wazuh/files/3794653/Scan-Build-Report-Agent.zip)

+ Manager report:
![image](https://user-images.githubusercontent.com/36703917/67963377-c5c81580-fbfe-11e9-976d-16b3e61f1019.png)

[Scan-Build-Report-Manager.zip](https://github.com/wazuh/wazuh/files/3794671/Scan-Build-Report-Manager.zip)

Best regards.

**Comment by chemamartinez** [comment]
Hi @sergiovp,

On CentOS7, with the CLANG 3.4.2, I got four bugs reported by `scan-build` for the 3.11 branch:

_Results_
```
scan-build: 4 bugs found.
scan-build: Run 'scan-view /tmp/scan-build-2019-11-04-012003-14461-1' to examine bug reports.
```
_clang version_
```
[root@centos7 src]# clang -v
clang version 3.4.2 (tags/RELEASE_34/dot2-final)
Target: x86_64-redhat-linux-gnu
Thread model: posix
Found candidate GCC installation: /bin/../lib/gcc/x86_64-redhat-linux/4.8.2
Found candidate GCC installation: /bin/../lib/gcc/x86_64-redhat-linux/4.8.5
Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/4.8.2
Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/4.8.5
Selected GCC installation: /bin/../lib/gcc/x86_64-redhat-linux/4.8.5
```
_gcc version_
```
[root@centos7 src]# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-39) (GCC)
```
_cmake version_
```
[root@centos7 src]# cmake -h
cmake version 2.8.12.2
```
![image](https://user-images.githubusercontent.com/29475387/68111726-22be1700-fef0-11e9-863c-84b897c4fc2e.png)

Could you paste here the output of the previous commands?

Regards.

**Comment by sergiovp** [comment]
Hello @chemamartinez , 

*Results*
~~~
scan-build: 38 bugs found.
scan-build: Run 'scan-view /tmp/scan-build-2019-11-04-100021-5040-1' to examine bug reports.
~~~

*clang version*
~~~
[root@ManagerCentOS7 vagrant]# clang -v
clang version 3.4.2 (tags/RELEASE_34/dot2-final)
Target: x86_64-redhat-linux-gnu
Thread model: posix
Found candidate GCC installation: /bin/../lib/gcc/x86_64-redhat-linux/4.8.2
Found candidate GCC installation: /bin/../lib/gcc/x86_64-redhat-linux/4.8.5
Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/4.8.2
Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/4.8.5
Selected GCC installation: /bin/../lib/gcc/x86_64-redhat-linux/4.8.5
~~~

*gcc version*
~~~
[root@ManagerCentOS7 vagrant]# gcc -v
Usando especificaciones internas.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Objetivo: x86_64-redhat-linux
Configurado con: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Modelo de hilos: posix
gcc versión 4.8.5 20150623 (Red Hat 4.8.5-39) (GCC)
~~~

*cmake version*
~~~
[root@ManagerCentOS7 bin]# cmake -h
cmake version 2.8.12.2
~~~

![image](https://user-images.githubusercontent.com/36703917/68114889-68caa900-fef7-11e9-99bb-9fd6f8187e56.png)


The steps I followed to run scan-build are
- make clean-internals
- scan-build make TARGET=server

Regards.

**Comment by Lopuiz** [comment]
Hello team,

Most of them are due to function `strdup` can return NULL if there isn't enough memory to reserve. It could be solved using `os_strdup` instead of `strdup`.

Respect `os_xml:52` bug, it's due following line:
`c = (_lxml->stash_i > 0) ? _lxml->stash[--_lxml->stash_i] : *(_lxml->string++);`
Scan-build report that `*(_lxml->string++)` can be null. But this function only will be called when value string isn't null.

Something similar happens in `wm_vuln_detector.c:4783` because `*cpe_s->part` never will be null.
Scan-build reports it because it can't see that variable named `part` can't be null:
```
CREATE TABLE IF NOT EXISTS CPE_INDEX (
	ID INTEGER,
    POS INTEGER,
    PART TEXT NOT NULL,
    VENDOR TEXT NOT NULL,
    PRODUCT TEXT NOT NULL,
    VERSION TEXT NOT NULL,
    UPDATEV TEXT,
    EDITION TEXT,
    LANGUAGE TEXT,
    SW_EDITION TEXT,
    TARGET_SW TEXT,
    TARGET_HW TEXT,
    OTHER TEXT,
    MSU_NAME TEXT,
    PRIMARY KEY(ID, POS)
);
```

Finally the bug found in `decode-xml.c:478` is due to following line:
`os_calloc(Config.decoder_order_size, sizeof(void *), pi->order);`
Although the `pi->order` type is incompatible with sizeof operand type `void *` the memory reserved are the same.

Best regards,
Eva

**Comment by chemamartinez** [comment]
Merged #4201
